ezmlm-checksub.1
.TH ezmlm-issubn 1
.SH NAME
ezmlm-checksub \- reject senders not subscribed to a mailing list
.SH SYNOPSIS
.B ezmlm-checksub
[
.B \-nN
]
.I dir
.I [ subdir1 ... ]
.SH DESCRIPTION
.B ezmlm-checksub
checks to see if the address obtained from the environment variable
.I SENDER
is subscribed to the mailing list stored in
.I dir/subdir1
or the mailing list in
.I dir/subdir2
or ...

If
.I SENDER
is not defined
.B ezmlm-checksub
exits with an error.

If
.B \-N
is set (the default) and
.I SENDER
is not on any of the mailing lists,
.B ezmlm-checksub
outputs the message named
.I TXT_ONLY_SUBSCRIBERS
and exits 100, indicating a permanent failure to qmail.

If
.B \-n
is set and
.I SENDER
is on any of the mailing lists,
.B ezmlm-checksub
outputs the message named
.I TXT_REJECT_POSTS
and exits 100.

Otherwise,
.B ezmlm-checksub
exits with a zero exit code (success).

If
.I subdir1
(or more) are present on the command line and are relative paths,
.B ezmlm-checksub
only examines the named subdirectories within the absolute path
.IR dir .

.B ezmlm-checksub
exits 111 on errors.
.SH OPTIONS
.TP
.B \-m \fIMESSAGE
Use
.I MESSAGE
in place of either the message named
.I TXT_ONLY_SUBSCRIBERS
or
.I TXT_REJECT_POSTS
in rejections.
.TP
.B \-n
Negative operation: ensure the sender is not a subscriber.  This is
useful when trying to exclude SENDERs.
.TP
.B \-N
(Default.)
Normal operation: ensure if the sender is a subscriber.
.SH "SEE ALSO"
ezmlm-issubn(1),
ezmlm-list(1),
ezmlm-manage(1),
ezmlm-make(1),
ezmlm-send(1),
ezmlm-sub(1),
ezmlm-unsub(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================


ezmlm-checksub.c
#include <unistd.h>
#include "strerr.h"
#include "env.h"
#include "str.h"
#include "subdb.h"
#include "subfd.h"
#include "substdio.h"
#include "sgetopt.h"
#include "messages.h"
#include "die.h"
#include "idx.h"
#include "config.h"
#include "auto_version.h"

const char FATAL[] = "ezmlm-checksub: fatal: ";
const char USAGE[] =
"ezmlm-checksub: usage: ezmlm-checksub [-m MESSAGE] [-nN] dir [subdir ...]";

void main(int argc,char **argv)
{
  const char *dir;
  const char *subdir;
  const char *addr;
  const char *message = 0;
  int flagsub = 1;
  int opt;
  int senderissub;

  addr = env_get("SENDER");
  if (!addr) die_sender();	/* REQUIRE sender */

  while ((opt = getopt(argc,argv,"m:nNvV")) != opteof)
    switch(opt) {
    case 'm': message = optarg; break;
    case 'n': flagsub = 0; break;
    case 'N': flagsub = 1; break;
    case 'v':
    case 'V': strerr_die2x(0, "ezmlm-checksub version: ",auto_version);
    default:
      die_usage();
    }

  startup(dir = argv[optind++]);
  initsub(0);
  if (message == 0)
    message = flagsub ? MSG(TXT_ONLY_SUBSCRIBERS) : MSG(TXT_REJECT_POSTS);

  if (optind >= argc)
    senderissub = !!issub(0,addr,0);
  else {
    senderissub = 0;
    while ((subdir = argv[optind++]) != 0) {
      if (issub(subdir,addr,0)) {
	senderissub = 1;
	break;
      }
    }
  }
  closesub();

  /* senderissub and flagsub are both either 0 or 1 */
  if (senderissub != flagsub) {
    substdio_puts(subfdout,message);
    substdio_putflush(subfdout,"\n",1);
    _exit(100);
  }
  _exit(0);
}
